File: impl/typescript/src/test/ts/cli.test.ts
import { run, type TextSink } from "../../main/ts/cli";
import { DEFAULT_DICTIONARY } from "../../main/ts/liff";

declare const console: { log(text: string): void };

class BufferSink implements TextSink {
    value = "";

    write(text: string): void {
        this.value += text;
    }
}

interface Result {
    readonly status: number;
    readonly stdout: string;
    readonly stderr: string;
}

function invoke(...arguments_: string[]): Result {
    const stdout = new BufferSink();
    const stderr = new BufferSink();
    const status = run(arguments_, stdout, stderr, DEFAULT_DICTIONARY, () => 0);
    return { status, stdout: stdout.value, stderr: stderr.value };
}

function check(condition: boolean, message: string): asserts condition {
    if (!condition) {
        throw new Error(message);
    }
}

function equal<T>(actual: T, expected: T, message: string): void {
    if (!Object.is(actual, expected)) {
        throw new Error(`${message}: got ${String(actual)}, want ${String(expected)}`);
    }
}

function randomFoundAndNotFound(): void {
    const random = invoke();
    equal(random.status, 0, "random status");
    check(random.stdout.startsWith("AASLEAGH\n"), "random output");
    equal(random.stderr, "", "random stderr");

    for (const query of ["banteer", "banteeer", "glutt", "bilb", "bil*"]) {
        const result = invoke(query);
        equal(result.status, 0, `found status for ${query}`);
        equal(result.stderr, "", `found stderr for ${query}`);
    }
    check(invoke("glutt").stdout.startsWith("GLUTT LODGE\n"), "glutt match");
    check(invoke("bilb").stdout.startsWith("BILBSTER\n"), "bilb match");
    check(invoke("bil*").stdout.startsWith("BILBSTER\n"), "unique glob match");
    check(invoke("symonds", "yat").stdout.startsWith("SYMOND'S YAT\n"),
        "joined query");

    const missing = invoke("xyzzy");
    equal(missing.status, 1, "not-found status");
    equal(missing.stdout, 'No definition found for "xyzzy".\n', "not-found output");
    equal(missing.stderr, "", "not-found stderr");
}

function suggestionsAndGlobBoundaries(): void {
    const ambiguous = invoke("high");
    equal(ambiguous.status, 1, "ambiguous status");
    equal(ambiguous.stdout,
        "Did you mean?\nHIGH LIMERIGG\nHIGH OFFLEY\nAITH\nCHICAGO\n",
        "ambiguous output");

    const many = invoke("b*");
    equal(many.status, 1, "large glob status");
    equal(many.stdout,
        "Did you mean?\n"
        + "BABWORTH\nBALDOCK\nBALLYCUMBER\nBANFF\nBANTEER\n"
        + "BARSTIBLEY\nBAUGHURST\nBAUMBER\nBEALINGS\nBEAULIEU HILL\n"
        + "and 44 others\n",
        "large glob output");

    const eleven = invoke("bo*");
    equal(eleven.status, 1, "eleven-result status");
    check(!eleven.stdout.includes("and "), "eleven results were truncated");
    equal(eleven.stdout.trimEnd().split("\n").length, 12, "eleven-result line count");

    const all = invoke("*");
    equal(all.status, 1, "all-result status");
    check(all.stdout.endsWith("and 540 others\n"), "all-result truncation");
    equal(all.stdout.trimEnd().split("\n").length, 12, "all-result line count");
}

function helpAndInvalidUsage(): void {
    const help = invoke("--help");
    equal(help.status, 0, "help status");
    check(help.stdout.startsWith("Usage: liff"), "help output");
    equal(help.stderr, "", "help stderr");

    const invalid = invoke("--unknown");
    equal(invalid.status, 2, "invalid status");
    equal(invalid.stdout, "", "invalid stdout");
    check(invalid.stderr.startsWith("Usage: liff"), "invalid stderr");
}

randomFoundAndNotFound();
suggestionsAndGlobBoundaries();
helpAndInvalidUsage();
console.log("cli.test: OK");

File: impl/lua/src/test/lua/cli_test.lua
local root = assert(arg[1], "repository root argument is required")
package.path = root .. "/impl/lua/src/main/lua/?.lua;"
    .. root .. "/impl/lua/src/main/lua/?/init.lua;"
    .. package.path

local cli = require("liff.cli")
local liff = require("liff")

local function sink()
    return {
        value = "",
        write = function(self, ...)
            for index = 1, select("#", ...) do
                self.value = self.value .. tostring(select(index, ...))
            end
        end,
    }
end

local function invoke(...)
    local arguments = { ... }
    local stdout = sink()
    local stderr = sink()
    local status = cli.run(arguments, stdout, stderr, liff.DEFAULT_DICTIONARY,
        function() return 0 end)
    return status, stdout.value, stderr.value
end

local function equal(actual, expected, message)
    if actual ~= expected then
        error(string.format("%s: got %q, want %q", message, tostring(actual), tostring(expected)))
    end
end

local function check(condition, message)
    if not condition then
        error(message)
    end
end

local status, output, errors = invoke()
equal(status, 0, "random status")
check(output:sub(1, #"AASLEAGH\n") == "AASLEAGH\n", "random output")
equal(errors, "", "random stderr")

for _, query in ipairs({ "banteer", "banteeer", "glutt", "bilb", "bil*" }) do
    status, output, errors = invoke(query)
    equal(status, 0, "found status")
    equal(errors, "", "found stderr")
end
_, output = invoke("glutt")
check(output:sub(1, #"GLUTT LODGE\n") == "GLUTT LODGE\n", "glutt match")
_, output = invoke("bilb")
check(output:sub(1, #"BILBSTER\n") == "BILBSTER\n", "bilb match")
_, output = invoke("symonds", "yat")
check(output:sub(1, #"SYMOND'S YAT\n") == "SYMOND'S YAT\n", "joined query")

status, output, errors = invoke("high")
equal(status, 1, "ambiguous status")
equal(output, "Did you mean?\nHIGH LIMERIGG\nHIGH OFFLEY\nAITH\nCHICAGO\n",
    "ambiguous output")

status, output = invoke("b*")
equal(status, 1, "large glob status")
equal(output,
    "Did you mean?\n"
        .. "BABWORTH\nBALDOCK\nBALLYCUMBER\nBANFF\nBANTEER\n"
        .. "BARSTIBLEY\nBAUGHURST\nBAUMBER\nBEALINGS\nBEAULIEU HILL\n"
        .. "and 44 others\n",
    "large glob output")

status, output = invoke("bo*")
equal(status, 1, "eleven-result status")
check(not output:find("and ", 1, true), "eleven results were truncated")
local _, eleven_lines = output:gsub("\n", "\n")
equal(eleven_lines, 12, "eleven-result line count")

status, output = invoke("*")
equal(status, 1, "all-result status")
check(output:sub(-#"and 540 others\n") == "and 540 others\n", "all-result truncation")

status, output, errors = invoke("xyzzy")
equal(status, 1, "not-found status")
equal(output, "No definition found for \"xyzzy\".\n", "not-found output")
equal(errors, "", "not-found stderr")

status, output, errors = invoke("--help")
equal(status, 0, "help status")
check(output:sub(1, #"Usage: liff") == "Usage: liff", "help output")
equal(errors, "", "help stderr")

status, output, errors = invoke("--unknown")
equal(status, 2, "invalid status")
equal(output, "", "invalid stdout")
check(errors:sub(1, #"Usage: liff") == "Usage: liff", "invalid stderr")

io.write("cli_test: OK\n")
